File: python/parmana/api/trust_record_api.py
"""
Parmana Trust Record API.

Retrieve Execution Trust Records.
"""

from __future__ import annotations

from urllib.parse import urlencode

from parmana.config.transport import Transport
from parmana.models.trust_record import ExecutionTrustRecord


class TrustRecordApi:
    """
    Trust Record API.

    Responsibilities
    ----------------
    - Retrieve Execution Trust Records
    - List Execution Trust Records

    This API does NOT:
    - execute Business Transactions
    - verify trust records
    - replay executions
    - generate receipts
    """

    def __init__(
        self,
        transport: Transport,
    ) -> None:
        self._transport = transport

    def get(
        self,
        business_transaction_id: str,
    ) -> ExecutionTrustRecord:
        """
        Retrieve an Execution Trust Record.

        Parameters
        ----------
        business_transaction_id:
            Business Transaction identifier.

        Returns
        -------
        Execution Trust Record.
        """

        return self._transport.send(
            method="GET",
            path=f"/trust-records/{business_transaction_id}",
            response_model=ExecutionTrustRecord,
        )

    def list(
        self,
        *,
        page: int = 1,
        page_size: int = 25,
        since: str | None = None,
        until: str | None = None,
    ) -> list[ExecutionTrustRecord]:
        """
        List Execution Trust Records, newest first.

        Maps to GET /trust-records. Only records of Business Transactions
        this caller submitted are returned, so a page may hold fewer than
        `page_size` records.

        Parameters
        ----------
        page:
            Page number.

        page_size:
            Number of Business Transactions per page.

        since:
            Only records of transactions created at or after this ISO 8601
            time.

        until:
            Only records of transactions created at or before this ISO 8601
            time.
        """

        query: dict[str, str] = {"page": str(page), "pageSize": str(page_size)}
        if since is not None:
            query["since"] = since
        if until is not None:
            query["until"] = until

        return self._transport.send(
            method="GET",
            path=f"/trust-records?{urlencode(query)}",
            response_model=list[ExecutionTrustRecord],
        )

File: typescript/src/client/TrustRecordApi.ts
/**
 * Parmana Trust Record API.
 *
 * Retrieve Execution Trust Records.
 */

import type { ExecutionTrustRecord } from "../models/index.js";

import type { Transport } from "../config/Transport.js";

/**
 * Trust Record API.
 */
export class TrustRecordApi {
  constructor(private readonly transport: Transport) {}

  /**
   * Retrieve an Execution Trust Record.
   */
  public async get(
    businessTransactionId: string,
  ): Promise<ExecutionTrustRecord> {
    const response = await this.transport.send<ExecutionTrustRecord>({
      method: "GET",
      path: `/trust-records/${businessTransactionId}`,
    });

    return response.body;
  }

  /**
   * List Execution Trust Records, newest first. Maps to GET /trust-records.
   * Only records of Business Transactions this caller submitted are
   * returned, so a page may hold fewer than `pageSize` records.
   *
   * @param options.since Only records of transactions created at or after
   *   this ISO 8601 time.
   * @param options.until Only records of transactions created at or before
   *   this ISO 8601 time.
   */
  public async list(
    page = 1,
    pageSize = 25,
    options: { readonly since?: string; readonly until?: string } = {},
  ): Promise<ExecutionTrustRecord[]> {
    const query = new URLSearchParams({
      page: String(page),
      pageSize: String(pageSize),
    });

    if (options.since !== undefined) query.set("since", options.since);
    if (options.until !== undefined) query.set("until", options.until);

    const response = await this.transport.send<ExecutionTrustRecord[]>({
      method: "GET",
      path: `/trust-records?${query.toString()}`,
    });

    return response.body;
  }
}
